internal/oci/oci.go

```diff
@@ -87,40 +87,61 @@ func Download(ctx context.Context, url string, target io.Writer, options ...Down
 	}
 
 	tag := imgref.Reference
-	if _, err := oras.Copy(ctx, repo, tag, fs, tag, oras.DefaultCopyOptions); err != nil {
-		return fmt.Errorf("failed to fetch artifact: %w", err)
+	desc, data, err := repo.Manifests().FetchReference(ctx, tag)
+	if err != nil {
+		return fmt.Errorf("failed to fetch manifest: %w", err)
 	}
+	defer data.Close()
 
-	files, err := os.ReadDir(tmpdir)
+	successors, err := content.Successors(ctx, repo, desc)
 	if err != nil {
-		return fmt.Errorf("failed to read temp dir: %w", err)
+		return fmt.Errorf("failed to fetch successors: %w", err)
 	}
 
-	// we always expect only one single file to be downloaded.
-	if len(files) == 0 {
-		return fmt.Errorf("no artifacts found")
-	} else if len(files) > 1 {
-		return fmt.Errorf("multiple artifacts found")
+	source, err := findArtifactDescriptor(successors, opts)
+	if err != nil {
+		return fmt.Errorf("failed to find artifact: %w", err)
 	}
 
-	fpath := filepath.Join(tmpdir, files[0].Name())
-	fp, err := os.Open(fpath)
+	// get a reader to the blob and copies it to the target.
+	reader, err := repo.Blobs().Fetch(ctx, source)
 	if err != nil {
-		return fmt.Errorf("failed to open file: %w", err)
+		return fmt.Errorf("failed to fetch blob: %w", err)
 	}
-	defer fp.Close()
+	defer reader.Close()
 
-	if _, err := io.Copy(target, fp); err != nil {
-		return fmt.Errorf("failed to copy file: %w", err)
+	if _, err := io.Copy(target, reader); err != nil {
+		return fmt.Errorf("failed to copy blob: %w", err)
 	}
 
 	return nil
 }
 
+// findArtifactDescriptor filters, out of the provided list of descriptors, the
+// one that matches the given options. If no artifact name is provided, it
+// returns the first descriptor.
+func findArtifactDescriptor(all []ocispec.Descriptor, opts downloadOptions) (ocispec.Descriptor, error) {
+	for _, desc := range all {
+		if desc.MediaType == ocispec.MediaTypeEmptyJSON {
+			continue
+		}
+		// if no artifact name is specified, we use the first one.
+		fname := opts.artifactName
+		if fname == "" || fname == desc.Annotations[ocispec.AnnotationTitle] {
+			return desc, nil
+		}
+	}
+	if opts.artifactName == "" {
+		return ocispec.Descriptor{}, fmt.Errorf("no artifacts found")
+	}
+	return ocispec.Descriptor{}, fmt.Errorf("artifact %q not found", opts.artifactName)
+}
```

Comment:
When having a look at our implementation, I realize that the ADR _just_ says: "We'll download via OCI", but not at all what's meant by this. As fart as I understand, there's no common understanding of what it means to download an artifact via OCI, i.e. it is very much implementation dependent.

I think we should augment the ADR with a spec what k0s means/expects when it tries to download files from OCI registries. WDYT?

Reply:
Oh yes, this is what we mention in the ADR:

> Implement support in Autopilot for pulling artifacts, such as k0s binaries and image bundles

So we at least know what we expect to find in the registry. We should at least write down what we expect when it comes to file names and formats.